File: examples/telemetry/14_data_governance.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""Data governance: consent levels, data classification, and redaction receipts.

Demonstrates:
  1. ConsentLevel — gate signal collection per user consent tier
  2. Data classification — label fields by sensitivity class; pair with PIIRules for enforcement
  3. RedactionReceipts — cryptographic audit trail for every PII redaction
"""

from __future__ import annotations

from provide.telemetry import (
    event,
    get_logger,
    register_pii_rule,
    setup_telemetry,
    shutdown_telemetry,
)
from provide.telemetry.classification import (
    ClassificationRule,
    DataClass,
    register_classification_rules,
)
from provide.telemetry.consent import ConsentLevel, set_consent_level, should_allow
from provide.telemetry.pii import PIIRule, sanitize_payload
from provide.telemetry.receipts import enable_receipts, get_emitted_receipts_for_tests


def demo_consent() -> None:
    print("── 1. Consent Levels ──────────────────────────────────────")
    for level in ConsentLevel:
        set_consent_level(level)
        logs_debug = should_allow("logs", "DEBUG")
        logs_error = should_allow("logs", "ERROR")
        traces = should_allow("traces")
        metrics = should_allow("metrics")
        ctx = should_allow("context")
        print(
            f"  {level.value:<12} "
            f"logs(DEBUG)={logs_debug!s:<5} "
            f"logs(ERROR)={logs_error!s:<5} "
            f"traces={traces!s:<5} "
            f"metrics={metrics!s:<5} "
            f"context={ctx}"
        )
    set_consent_level(ConsentLevel.FULL)
    print()


def demo_classification() -> None:
    print("── 2. Data Classification ─────────────────────────────────")
    # Register rules: pattern → DataClass label
    register_classification_rules(
        [
            ClassificationRule(pattern="ssn", classification=DataClass.PII),
            ClassificationRule(pattern="card_number", classification=DataClass.PCI),
            ClassificationRule(pattern="diagnosis", classification=DataClass.PHI),
            ClassificationRule(pattern="api_*", classification=DataClass.SECRET),
        ]
    )
    # Classification adds __key__class labels to sanitized output.
    # Enforcement (drop, hash, redact) is applied by registering PIIRules per class.
    register_pii_rule(PIIRule(path=("ssn",), mode="redact"))
    register_pii_rule(PIIRule(path=("card_number",), mode="hash"))
    register_pii_rule(PIIRule(path=("diagnosis",), mode="drop"))
    register_pii_rule(PIIRule(path=("api_key",), mode="drop"))

    payload = {
        "user": "alice",
        "ssn": "123-45-6789",
        "card_number": "4111111111111111",
        "diagnosis": "hypertension",
        "api_key": "sk-prod-abc123",  # pragma: allowlist secret
    }
    cleaned = sanitize_payload(payload, enabled=True)

    print("  Field values after sanitization:")
    for k in payload:
        print(f"    {k}: {cleaned.get(k, '<dropped>')!r}")

    print("\n  Classification labels added to output:")
    for k, v in cleaned.items():
        if k.endswith("__class"):
            print(f"    {k}: {v!r}")
    print()


def demo_receipts() -> None:
    print("── 3. Redaction Receipts ──────────────────────────────────")
    # Enable test-mode receipt collection (in production, receipts are logged/forwarded)
    from provide.telemetry import receipts as _r

    _r._reset_receipts_for_tests()  # type: ignore[attr-defined]
    _r._test_mode = True  # type: ignore[attr-defined]
    enable_receipts(
        enabled=True,
        signing_key="demo-hmac-key",  # pragma: allowlist secret
        service_name="governance-demo",
    )

    register_pii_rule(PIIRule(path=("password",), mode="redact"))
    sanitize_payload({"user": "bob", "password": "s3cr3t"}, enabled=True)  # pragma: allowlist secret

    receipts = get_emitted_receipts_for_tests()
    if receipts:
        r = receipts[-1]
        print(f"  receipt_id:    {r.receipt_id}")
        print(f"  field_path:    {r.field_path}")
        print(f"  action:        {r.action}")
        print(f"  original_hash: {r.original_hash[:16]}...")
        print(f"  hmac:          {r.hmac[:16]}..." if r.hmac else "  hmac:          (unsigned)")
    enable_receipts(enabled=False)
    print()


def main() -> None:
    setup_telemetry()
    log = get_logger("governance-demo")
    log.info(event("governance", "demo", "start"))

    print("=== Data Governance Demo ===\n")
    demo_consent()
    demo_classification()
    demo_receipts()

    shutdown_telemetry()
    print("=== Done ===")


if __name__ == "__main__":
    main()

File: typescript/examples/telemetry/14_data_governance.ts
#!/usr/bin/env npx tsx
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * 14_data_governance — consent levels, data classification, and redaction receipts.
 *
 * Demonstrates:
 *  - ConsentLevel: gate signal collection per user consent tier
 *  - Data classification: label fields by sensitivity class; pair with PIIRules for enforcement
 *  - RedactionReceipts: cryptographic audit trail for every PII redaction
 *
 * Run: npx tsx examples/telemetry/14_data_governance.ts
 */

import {
  event,
  getLogger,
  registerPiiRule,
  sanitizePayload,
  setupTelemetry,
  shutdownTelemetry,
} from '@provide-io/telemetry';
import { setConsentLevel, shouldAllow } from '@provide-io/telemetry';
import { registerClassificationRules } from '@provide-io/telemetry';
import {
  enableReceipts,
  getEmittedReceiptsForTests,
  resetReceiptsForTests,
} from '@provide-io/telemetry';

// ── 1. Consent Levels ────────────────────────────────────────────────────────
function demoConsent(): void {
  console.log('── 1. Consent Levels ──────────────────────────────────────');
  const levels = ['FULL', 'FUNCTIONAL', 'MINIMAL', 'NONE'] as const;
  for (const level of levels) {
    setConsentLevel(level);
    const logsDebug = shouldAllow('logs', 'DEBUG');
    const logsError = shouldAllow('logs', 'ERROR');
    const traces = shouldAllow('traces');
    const metrics = shouldAllow('metrics');
    const ctx = shouldAllow('context');
    console.log(
      `  ${level.padEnd(12)} logs(DEBUG)=${String(logsDebug).padEnd(5)} ` +
        `logs(ERROR)=${String(logsError).padEnd(5)} traces=${String(traces).padEnd(5)} ` +
        `metrics=${String(metrics).padEnd(5)} context=${ctx}`,
    );
  }
  setConsentLevel('FULL');
  console.log();
}

// ── 2. Data Classification ───────────────────────────────────────────────────
function demoClassification(): void {
  console.log('── 2. Data Classification ─────────────────────────────────');
  // Register rules: pattern → DataClass label
  registerClassificationRules([
    { pattern: 'ssn', classification: 'PII' },
    { pattern: 'card_number', classification: 'PCI' },
    { pattern: 'diagnosis', classification: 'PHI' },
    { pattern: 'api_*', classification: 'SECRET' },
  ]);
  // Classification adds __key__class labels to sanitized output.
  // Enforcement (drop, hash, redact) is applied by registering PIIRules per class.
  registerPiiRule({ path: 'ssn', mode: 'redact' });
  registerPiiRule({ path: 'card_number', mode: 'hash' });
  registerPiiRule({ path: 'diagnosis', mode: 'drop' });
  registerPiiRule({ path: 'api_key', mode: 'drop' });

  // sanitizePayload mutates in-place; capture originals before sanitizing
  const originals: Record<string, unknown> = {
    user: 'alice',
    ssn: '123-45-6789',
    card_number: '4111111111111111',
    diagnosis: 'hypertension',
    api_key: 'sk-prod-abc123', // pragma: allowlist secret
  };
  const payload: Record<string, unknown> = { ...originals };
  sanitizePayload(payload);

  console.log('  Field values after sanitization:');
  for (const k of Object.keys(originals)) {
    const out = k in payload ? JSON.stringify(payload[k]) : '<dropped>';
    console.log(`    ${k}: ${out}`);
  }

  console.log('\n  Classification labels added to output:');
  for (const [k, v] of Object.entries(payload)) {
    if (k.endsWith('__class')) {
      console.log(`    ${k}: ${JSON.stringify(v)}`);
    }
  }
  console.log();
}

// ── 3. Redaction Receipts ────────────────────────────────────────────────────
function demoReceipts(): void {
  console.log('── 3. Redaction Receipts ──────────────────────────────────');
  resetReceiptsForTests();
  enableReceipts({
    enabled: true,
    signingKey: 'demo-hmac-key', // pragma: allowlist secret
    serviceName: 'governance-demo',
  });

  registerPiiRule({ path: 'password', mode: 'redact' });
  const receiptPayload = { user: 'bob', password: 's3cr3t' }; // pragma: allowlist secret
  sanitizePayload(receiptPayload);

  const receipts = getEmittedReceiptsForTests();
  if (receipts.length > 0) {
    const r = receipts[receipts.length - 1];
    console.log(`  receiptId:    ${r.receiptId}`);
    console.log(`  fieldPath:    ${r.fieldPath}`);
    console.log(`  action:       ${r.action}`);
    console.log(`  originalHash: ${r.originalHash.slice(0, 16)}...`);
    console.log(r.hmac ? `  hmac:         ${r.hmac.slice(0, 16)}...` : '  hmac:         (unsigned)');
  } else {
    console.log('  (no receipts captured)');
  }
  enableReceipts({ enabled: false });
  console.log();
}

// ── Main ─────────────────────────────────────────────────────────────────────
async function main(): Promise<void> {
  setupTelemetry({ serviceName: 'governance-demo' });
  const log = getLogger('governance-demo');
  log.info({ ...event('governance', 'demo', 'start') });

  console.log('=== Data Governance Demo ===\n');
  demoConsent();
  demoClassification();
  demoReceipts();

  await shutdownTelemetry();
  console.log('=== Done ===');
}

main().catch(console.error);
